Add consolidated build factory generator for the LLVM unified tree builds.

vvereschaka opened the pull request:
Added new build factory UnifiedTreeBuilder.getCmakeExBuildFactory().
The factory supports the latest Buildbot's CMake command build steps and
its format of arguments.

This factory is a replacement for the following factories:
* UnifiedTreeBuilder.getCmakeBuildFactory()
* UnifiedTreeBuilder.getCmakeWithNinjaBuildFactory()
* UnifiedTreeBuilder.getCmakeWithNinjaWithMSVCBuildFactory()
* XToolchainBuilder.getCmakeWithMSVCBuildFactory()

vvereschaka commented:
@gkistanova just in case, if you think we need someone from the community to review these changes, please invite them.

slydiman commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -559,3 +562,393 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+    """ Create and configure a builder factory to build a LLVM project from the unified source tree.
+
+        The is the CMake configurable build that uses Ninja generator by default. Using the other CMake generators
```
A typo?

vvereschaka replied:
fixed

slydiman commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -559,3 +562,393 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+            (see LLVMBuildFactory for more details).
+
+        cmake_definitions : dect, optional
```
dect -> dict

vvereschaka replied:
fixed

slydiman commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -559,3 +562,393 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        checks : list, optional
+            A list of the check targets (default is ["check-all"]).
+
+            Each check target gets executed with own step. The check step use LitTestCommand
```
use -> uses

vvereschaka replied:
fixed

slydiman commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
````diff
@@ -559,3 +562,393 @@ def getCmakeWithNinjaMultistageBuildFactory(
+            ]
+            ```
+
+        generator : str, required
````
I think this parameter is optional due to a default value. A default value does not make sense if this parameter is required. Maybe specify `cannot be None`?

vvereschaka replied:
™cannot be None" means "required". I think "required" is ok to be here

slydiman commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -559,3 +562,393 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+            f.addStep(
+                steps.CMake(
+                    name            = f"build-{target_title:.42}",
```
Is the name length 48 restricted somewhere?

vvereschaka replied:
This is a DB restriction. Max length is 50 characters for the step name. 48 + 2 to count the identical names.

vvereschaka commented:
Fixed the typos.
@gkistanova @andreil99 would you take a look?

vvereschaka commented:
fixed `prop:obj_dir` => `prop:objdir`

slydiman reviewed (approved):
LGTM

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,448 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        cmake_definitions = None,
+        cmake_options = None,
+        allow_cmake_defaults = True,    # Add default CMake definitions to build LLVM project (if not specified).
+        targets = ["."],
```
Please do not use mutable objects as defaults. This leads to hard-to-catch bugs. Instead, set it to `None` and handle that in the code by setting up the default value if `None`.

vvereschaka replied:
Sorry, I need a usage of these default values to support a intended logic. I don't see any reasons to do not use them since I don't change them within the function. This is an useful feature of Python.

gkistanova replied:
What's wrong with
```
targets = None,
...
if targets is None:
    targets = ["."]
...
```
?
Empty list would mean no targets to build.

vvereschaka replied:
In regular meaning a passing of None supposes doing nothing; or don't use this argument. A passing of something means doing something with this parameter value, even if it's an empty list (probably some default behaviour). This is an optimal meaning I think. Sure, we can change it and a passing of None will mean "use default value", but a passing of an empty list will mean "do nothing". But actually I don't see a good reason for it.

Maybe I could remove all default targets for all build operations such as check-all & etc. and force people to explicitly specify the targets everywhere to allow the step, but I would prefer to enable the default build by default and let people explicitly disable this step if necessary (by passing None).
.

gkistanova replied:
Make it immutable then. Tuple, for example
```
targets = ("."),
```
would do what you want without following a python anti-pattern.

vvereschaka replied:
I used the strings instead of the lists there.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,448 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        cmake_options = None,
+        allow_cmake_defaults = True,    # Add default CMake definitions to build LLVM project (if not specified).
+        targets = ["."],
+        install_targets = ["install"],
```
ditto

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,448 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        allow_cmake_defaults = True,    # Add default CMake definitions to build LLVM project (if not specified).
+        targets = ["."],
+        install_targets = ["install"],
+        checks = ["check-all"],
```
ditto

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,448 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        if not "LLVM_ENABLE_ASSERTIONS" in cmake_definitions:
+            cmake_definitions.update({ "LLVM_ENABLE_ASSERTIONS" : "ON" })
+        if not "LLVM_LIT_ARGS" in cmake_definitions and checks:
+            cmake_definitions.update({ "LLVM_LIT_ARGS" : "-v -vv" })
```
`-v` implies `-vv`. Specifying the both is redundant. Nice to fix but is not a blocker for this patch, though.

vvereschaka replied:
Updated accordingly

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,448 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+            f.addStep(
+                steps.CMake(
+                    name            = f"build-{target_title:.42}",
```
We have the step name length increased to 255. See https://github.com/gkistanova/buildbot/commit/b745a335ebdbb58ec2f09b7e87242b72f2a3b0bc. No real need to truncate it to 42.

vvereschaka replied:
The buildbot does not support the long step names by itself. The local copy of the buildbot will get unpredictable failed in case of specifying of the long step names I suppose.
Also, the long step names break the web ui. I think it is better to cut the step names to the max defaults.

gkistanova replied:
I'm fine with truncating the names, I'm not easy with hardcoded lengths, though. It will be a wrong length next time somebody will rename the step. But we can keep it as is if you feel strongly about that.

vvereschaka replied:
Not a big deal actually. Now I think that's not my business to worry about the length of the step names. This is the buildbot's responsibility if there are any limits there.

gkistanova replied:
Exactly my point.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,448 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        for target in checks:
+            f.addStep(
+                LitTestCommand(
+                    name            = f"test-{target:.43}",
```
ditto.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,448 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        for check, cmd in checks_on_target:
+            f.addStep(
+                LitTestCommand(
+                    name            = f"test-{check:.43}",
```
ditto

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,448 @@ def getCmakeWithNinjaMultistageBuildFactory(
+            target_title = target if target.startswith("install") else f"install-{target}"
+            f.addStep(
+                steps.CMake(
+                    name            = f"{target_title:.48}",
```
ditto.

gkistanova reviewed (changes requested):
Thanks, Vlad! The patch looks good with a few nit picks.

vvereschaka committed:
Add consolidated build factory generator for the LLVM unified tree builds.

Added new build factory UnifiedTreeBuilder.getCmakeExBuildFactory().
The factory supports the latest Buildbot's CMake command build steps and
its format of arguments.

This factory is a replacement for the following factories:
* UnifiedTreeBuilder.getCmakeBuildFactory()
* UnifiedTreeBuilder.getCmakeWithNinjaBuildFactory()
* UnifiedTreeBuilder.getCmakeWithNinjaWithMSVCBuildFactory()
* XToolchainBuilder.getCmakeWithMSVCBuildFactory()

vvereschaka commented:
@gkistanova @andreil99 

The latest updates for PR:
* used strings instead of the lists as the default values for the factory arguments;
* normalized the list arguments with the target specifications;
* removed the step name truncations;
* updated the default LIT test parameters: removed `-vv`, added `--time-tests`;
* tried to handle `util.Interpolate` properly at some places;
* added initial tests for `UnifiedTreeBuilder.getCmakeExBuildFactory`;

I'll add more tests. I have converted this PR to `Draft` for awhile until that moment.

vvereschaka commented:
>I'll add more tests. I have converted this PR to Draft for awhile until that moment.

I have finished with the tests for this factory. PR is ready to review.

vvereschaka commented:
I run the test something like that

```
(venv3) <llvm_project_root>\sandbox\llvm-zorg-3.x\test\buildbot>python <llvm_project_root>\llvm-project\llvm\utils\lit\lit.py -v .
-- Testing: 1 tests, 1 workers --
PASS: Zorg :: buildbot/builders/unified_cmakeex.py (1 of 1)

Testing Time: 2.29s

Total Discovered Tests: 1
  Passed: 1 (100.00%)
```

slydiman commented on test/buildbot/builders/unified_cmakeex.py:
```diff
@@ -0,0 +1,352 @@
+    )
+print(f"Step Extendions: {f}")
+
+assert factory_has_num_steps(f, 9 + 5)
```
Why `9+5`? I expected some preconfigured steps +4 added explicitly.

vvereschaka replied:
Correct, preconfigured + 5 added explicity

slydiman reviewed (approved):
LGTM generally.

I just have one cosmetic question.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+            If "auto" is specified, the runtime projects will be extracted from 'depends_on_projects' parameter.
+
+            If None is specified, LLVM_ENABLE_RUNTIMES will avoided for the CMake configuration step.
```
A mistype, it seems - "will avoided"? How about something like "LLVM_ENABLE_RUNTIMES will not be set" instead?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        allow_cmake_defaults : boolean
+            Add default CMake definitions to the build configuration step (default True).
+
+            The default value will be avoided if definition is already specified with the 'cmake_definitions' argument.
```
How about "A default value will be used only if a definition is not explicitly specified in 'cmake_definitions' argument and `allow_cmake_defaults` evaluates to `True`"?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+            The default value will be avoided if definition is already specified with the 'cmake_definitions' argument.
+
+            The following default CMake definitions will be added:
```
This is nice to have documented, though, unfortunately a "re-phrase of the code below" kind of comments are getting easily outdated and out of sync with the code. If you strongly feel for getting this documented, could you add a comment "If you are changing the code here please do not forget to update the documentation part above." in the line 239, please?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        targets : list, optional
+            A list of targets to build (default is ["."]).
+
+            Pass '.' in a list to build the default target. Each target gets built with own step.
```
"Each target gets built in a separate step, skipped if None"?

It seems the first sentence does not related to the second one to be a one paragraph? The next NOTE tells about default targets.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+            Pass '.' in a list to build the default target. Each target gets built with own step.
+
+            NOTE: The default target (.) cannot be specified by the IRenderable object.
```
"NOTE: The default target (.) must be specified as a string."? Or what else it could be?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        install_targets : list, optional
+            A list of the installation targets (default is ["install"]).
+
+            Each target gets installed with own step.
```
"with own step" -> "in a separate step" here and everywhere else in this file.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+            Each target gets installed with own step.
+
+            NOTE: the 'install_dir' argument also must be provided to add the installation steps.
```
"NOTE: 'install_dir' argument must be specified if there are install_targets."?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        checks_on_target : list of tuples, optional
+            A list of commands to run the tests on the remote dev boards (default is None).
+
+            This argument expects a list of tuples, where is a name of the test (to display) and
```
"list of tuples, where is a name of the test (to display) and the test command to execute:" -> "list of tuples with test names to display and the test commands to execute. For example:"?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        vs_arch : str, optional
+            Provide a build host arch for the Visual Studio tools (default is None)
+
+            Possible values are None, "amd64", "x64" & etc.
```
`None, "amd64", "x64" & etc.` -> `None, "amd64", "x64", etc. Please see Visual Studio documentation for more details.`

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        extra_git_args : dict, optional
+            Provide extra arguments for the Git step (default is None).
+
+            Sometimes is necessary to pass some additional parameters into the git step, such as
```
"Sometimes is necessary" -> "Sometimes it is necessary"?
"into the git step" -> "to a  git step"?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+            Provide extra arguments for the Git step (default is None).
+
+            Sometimes is necessary to pass some additional parameters into the git step, such as
+            'config', 'workdir' & etc.
```
"& etc" -> ", etc." here and everywhere else in this file.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+            A specific root project directory within the LLVM source code to configure and build (default is "llvm").
+
+            Provide a relative path to sub-project within the llvm-project directory to start configuring with instead of
+            default llvm folder.
```
This sentence is redundant to the previous one and could be removed without loosing much.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+            Provide a relative path to sub-project within the llvm-project directory to start configuring with instead of
+            default llvm folder.
+
+            As example: pass "flang/runtime" to configure and build the Flang runtime sub-project.
```
"As example:" -> "For example,"?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        pre_configure_steps : list or LLVMFactory, optional
+            The buildflow customization steps to execute before the CMake configuration step.
+
+            Provide a list of the build step objects or LLVMFactory object.
```
This sentence is redundant to the previous one and could be removed without loosing much.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        post_build_steps : list or LLVMFactory, optional
+            The buildflow customization steps to execute after the build step, but before the check steps.
+
+            Provide a list of the build step objects or LLVMFactory object.
```
"the build step objects" -> "build step objects"?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        pre_install_steps : list or LLVMFactory, optional
+            The buildflow customization steps to execute after the check steps, but before the installation steps (if requested).
+
+            Provide a list of the build step objects or LLVMFactory object.
```
This sentence is redundant to the previous one and could be removed without loosing much.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        post_finalize_steps : list or LLVMFactory, optional
+            The buildflow customization steps to execute at the end of the build workflow.
+
+            Provide a list of the build step objects or LLVMFactory object.
```
This sentence is redundant to the previous one and could be removed without loosing much.

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        srcdir
+            Relative path to the source code root directory ("llvm-project" by default).
+
+            "%(prop:builddir)s/%(prop:srcdir)s" is a full path to the source code dir.
```
"a full path" -> "a fully qualified path"?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+            "%(prop:builddir)s/%(prop:srcdir)s" is a full path to the source code dir.
+
+        srcdir_relative
+            Relative path from the obj dir to the source code root dir ("../llvm-project" by default).
```
"Path to the source code root directory relative to the `objdir` ("../llvm-project" by default)."?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        objdir
+            Relative path to the build directory.
+
+            "%(prop:builddir)s/%(prop:objdir)s" is a full path to the obj dir.
```
"a full path" -> "a fully qualified path"?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+            "%(prop:builddir)s/%(prop:objdir)s" is a full path to the obj dir.
+
+        depends_on_projects
+            A list of the depended projects for the build.
```
"A list of LLVM projects this builder depends on. It will build commits to these projects."?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+            A list of enabled runtimes.
+
+    """
+    assert generator, "The CMake generator must be specified."
```
"The CMake generator" -> "CMake generator"?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+    f.addSteps([
+        # Set up some properties, which could be used to configure the builders.
+        steps.SetProperties(
```
Could you elaborate why do you need build properties for these, please?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+
+    f.addGetSourcecodeSteps(**extra_git_args)
+
+    # Add the custom pre-configuration workflow extension steps, if specified.
```
"Add custom pre-configuration steps, if specified."?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+    if pre_configure_steps:
+        extend_with_custom_steps(f, pre_configure_steps)
+
+    # Configure MSVC environment at first if requested.
```
"Configure MSVC environment if requested."?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+    if vs:
+        f.addStep(
+            steps.SetPropertyFromCommand(
+                name            = "set-pros.vs_env",
```
"pros" -> "props"?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+    assert isinstance(cmake_options, list), "The CMake options argument must be a list."
+
+    if not "LLVM_ENABLE_PROJECTS" in cmake_definitions and f.enable_projects:
+        cmake_definitions.update({ "LLVM_ENABLE_PROJECTS" : ";".join(sorted(f.enable_projects)) })
```
Why do you need to sort alphabetically the list of enabled projects?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        cmake_definitions.update({ "LLVM_ENABLE_PROJECTS" : ";".join(sorted(f.enable_projects)) })
+
+    if not "LLVM_ENABLE_RUNTIMES" in cmake_definitions and f.enable_runtimes:
+        cmake_definitions.update({ "LLVM_ENABLE_RUNTIMES" : ";".join(sorted(f.enable_runtimes)) })
```
Why do you need to sort alphabetically the list of enabled runtimes?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        ))
+
+    # Build Commands.
+    #NOTE: please note that the default target (.) cannot be specified by the IRenderable object.
```
Please see my comment in the doc part. It seems the default target must be a string, nothing else, right?

gkistanova commented on zorg/buildbot/builders/UnifiedTreeBuilder.py:
```diff
@@ -567,3 +571,463 @@ def getCmakeWithNinjaMultistageBuildFactory(
+        ),
+    ])
+
+    # Get the source code steps at first. We share it between all stages.
```
"Let's start from getting the source code.  We share it between all stages."?

gkistanova reviewed (commented):
Thanks, Vlad.
The patch looks good. I have a few question and offered some language polishing here and there.

gkistanova reviewed (approved):
Let's move on with this patch, and see it in action.
